### PR TITLE
implementation of inlined operation input / output shapes

### PR DESCRIPTION
**Pull request by JordonPhillips**
This is an implementation of: #962

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java [hunk 1]
```diff
@@ -198,6 +202,10 @@ private void parseControlSection() {
 
             if (key.equals("version")) {
                 onVersion(value);
+            } else if (key.equals("operationInputSuffix")) {
+                operationInputSuffix = value.expectStringNode().getValue();
```
Comment (mtdowling) [inline smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java:206]: How much do we care about dupes in a single file? Seems like it would be silly for someone to duplicate these keys right, since they have to be at the top of the file... so guarding against it is probably going overboard.

Guarding against multiple versions per/file is still reasonable IMO because it's going to be used with low-level tools like grep that don't know Smithy model semantics. But these inline things probably won't be.

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlNodeParser.java [hunk 2]
```diff
@@ -135,6 +135,10 @@ static ObjectNode parseObjectNode(IdlModelParser parser, String parent) {
                 String key = parseNodeObjectKey(parser);
                 parser.ws();
                 parser.expect(':');
+                if (parser.peek() == '=') {
+                    throw parser.syntax("Defining structures inline with the `:=` syntax may only be used when "
```
Comment (mtdowling) [inline smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlNodeParser.java:139]: This doesn't seem accurate. We're not defining a structure here, it's an object node.

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java [hunk 3]
```diff
@@ -570,37 +584,120 @@ private void parseStructuredShape(
...
+                case "input":
+                    parseInlineableOperationMember(id, operationInputSuffix, builder::input);
+                    break;
+                case "output":
+                    parseInlineableOperationMember(id, operationOutputSuffix, builder::output);
+                    break;
+                case "errors":
+                    parseIdList(builder::addError);
+                    break;
+                default:
```
Comment (mtdowling) [inline smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java:627]: Could we ditch the OPERATION_PROPERTY_NAMES stuff and the validation in parseProperties, and instead handle that here in the switch? It seems redundant.

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java [hunk 4]
```diff
@@ -570,37 +584,120 @@ private void parseStructuredShape(
...
+        });
+        modelFile.onShape(builder);
+    }
+
+    private void parseProperties(ShapeId id, Collection<String> allowedProperties, Consumer<String> valueParser) {
+        ws();
+        expect('{');
+        ws();
+
+        Set<String> defined = new TreeSet<>();
```
Comment (mtdowling) [inline smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java:639]: Was there a reason for making this a TreeSet instead of a HashMap?

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java [hunk 5]
```diff
@@ -570,37 +584,120 @@ private void parseStructuredShape(
...
+                    parseIdList(builder::addError);
+                    break;
+                default:
+                    break;
+            }
+        });
+        modelFile.onShape(builder);
+    }
+
+    private void parseProperties(ShapeId id, Collection<String> allowedProperties, Consumer<String> valueParser) {
```
Comment (mtdowling) [inline smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java:634]: I like this abstraction!

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java [hunk 6]
```diff
@@ -570,37 +584,120 @@ private void parseStructuredShape(
...
+        expect('}');
+    }
+
+    private void parseInlineableOperationMember(ShapeId id, String suffix, Consumer<ShapeId> consumer) {
+        if (peek() == '=') {
+            if (!modelFile.getVersion().supportsInlineOperationIO()) {
+                throw syntax(id, "Inlined operation inputs and outputs can only be used with Smithy version 2 or "
+                        + "later. Attempted to use inlined IO with version `" + modelFile.getVersion() + "`.");
+            }
+            expect('=');
```
Comment (mtdowling) [inline smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java:664]: I think you need to clear pending docs here. For example, what happens right now with this?

```
operation Foo {
    input 
    /// These should not be part of the shape!
    := {}
}
```

**Review by mtdowling — CHANGES_REQUESTED** [review]

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java (hunk 1 above)
Comment (JordonPhillips) [reply smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java:206]: I don't think it matters that much to be honest, but I can add some generic check to make sure no control statement is duplicated.

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlNodeParser.java (hunk 2 above)
Comment (JordonPhillips) [reply smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlNodeParser.java:139]: I updated the wording to be more clear

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java (hunk 4 above)
Comment (JordonPhillips) [reply smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java:639]: At one point I had all the defined keys as part of the error message, but there's no point anymore

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlNodeParser.java [hunk 7]
```diff
@@ -159,7 +163,7 @@ static String parseNodeObjectKey(IdlModelParser parser) {
         }
     }
 
-    private static ArrayNode parseArrayNode(IdlModelParser parser) {
+    static ArrayNode parseArrayNode(IdlModelParser parser) {
```
Comment (mtdowling) [inline smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlNodeParser.java:166]: Is this change needed? Or did I miss the new class where this is being used?

File: smithy-model/src/main/java/software/amazon/smithy/model/shapes/SmithyIdlModelSerializer.java [hunk 8]
```diff
@@ -145,15 +145,39 @@ private String serialize(Model fullModel, Collection<Shape> shapes) {
...
         shapes.stream()
                 .filter(FunctionalUtils.not(Shape::isMemberShape))
+                .filter(shape -> !inlineableShapes.contains(shape.getId()))
                 .sorted(new ShapeComparator())
                 .forEach(shape -> shape.accept(shapeSerializer));
 
         return serializeHeader(fullModel, namespace) + codeWriter.toString();
     }
 
+    private Set<ShapeId> getInlineableShapes(Model fullModel, Collection<Shape> shapes) {
```
Comment (mtdowling) [inline smithy-model/src/main/java/software/amazon/smithy/model/shapes/SmithyIdlModelSerializer.java:160]: We'll need to modify this once #980 lands

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlNodeParser.java (hunk 7 above)
Comment (JordonPhillips) [reply smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlNodeParser.java]: I was using it at one point for the operation errors, but I ended up removing it. It's changed back now.

File: smithy-model/src/main/java/software/amazon/smithy/model/shapes/SmithyIdlModelSerializer.java (hunk 8 above)
Comment (JordonPhillips) [reply smithy-model/src/main/java/software/amazon/smithy/model/shapes/SmithyIdlModelSerializer.java:160]: Yeah, we'll also need to add the input and output shapes by default, so we can change it as part of that.

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java [hunk 9]
```diff
@@ -575,34 +583,89 @@ private void parseStructuredShape(
...
+            ws();
+        } while (peek() != '{');
     }
 
     private void parseOperationStatement(ShapeId id, SourceLocation location) {
         ws();
         OperationShape.Builder builder = OperationShape.builder().id(id).source(location);
-        ObjectNode node = IdlNodeParser.parseObjectNode(this, id.toString());
+
+        ObjectNode node = IdlNodeParser.parseObjectNode(this, id.toString(), (parser, key) -> {
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java]: Let's do service and resource in a seperate PR.

**Review by mtdowling — APPROVED** [review]

**Review by kstich — APPROVED** [review]
